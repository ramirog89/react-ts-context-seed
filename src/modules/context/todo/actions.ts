export enum ActionType {
  ADD_TODO = '[todo] add todo',
  REMOVE_TODO = '[todo] remove todo'
}
