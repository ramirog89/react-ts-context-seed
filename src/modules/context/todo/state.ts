import { TodoModel } from '../../../models';

export interface IState {
  todoList: TodoModel.ITodo[];
};

export const initialState: IState = {
  todoList: [
    {
      id: 1,
      title: 'post 1',
      description: 'this is a description',
      status: false,
    },
    {
      id: 2,
      title: 'post 2',
      description: 'this is a description 2',
      status: true,
    },
  ]
}
