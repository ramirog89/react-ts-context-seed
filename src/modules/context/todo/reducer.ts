import { ActionType } from './actions';
import { IState } from './state';
import { GeneralModel } from '../../models';

export const reducer = (state: IState, action: GeneralModel.IAction<ActionType>) => {
  switch (action.type) {
    case ActionType.ADD_TODO: {
      return { todoList: [...state.todoList, action.payload.todo] };
    }
    case ActionType.REMOVE_TODO: {
      return { todoList: state.todoList.filter(todo => action.payload.id !== todo.id) };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
};

