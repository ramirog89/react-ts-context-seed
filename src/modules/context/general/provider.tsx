import { useMemo, createContext, useReducer, useContext } from 'react';

import { ENV } from '../../../constants';
import { GeneralModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';

export const GeneralContext = createContext<
  {
    state: IState;
    addToast: (todo: GeneralModel.IToast) => void;
  } | null
>(null);

export const useGeneral = () => {
  const context = useContext(GeneralContext);
  if (!context) {
    throw new Error('GeneralContext is not wrapped with GeneralContext.Provider');
  }
  return context;
};

export const GeneralProvider = (props: IProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addToast = (toast: GeneralModel.IToast) => {
    const id = new Date().toString();
    dispatch({ type: ActionType.ADD_TOAST , payload: { toast: { ...toast, id } } });
    setTimeout(() => {}, ENV.TOAST_DELAY);
    dispatch({ type: ActionType.REMOVE_TOAST, payload: { id } });
  };

  const value = useMemo(() => ({ state, addToast }), [state]);

  return <GeneralContext.Provider value={value} {...props} />
};

