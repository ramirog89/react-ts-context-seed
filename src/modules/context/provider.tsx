import React from 'react';
import { ApiService } from '../services/ApiService';

import { ComposeProvider, Provider } from './compose';
import { IDependencies } from './rootState';
import { TodoProvider } from './todo/provider';

export const dependencies: IDependencies = {
  apiService: new ApiService(),
};

export const providers: Provider[] = [TodoProvider];

export const ContextProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ComposeProvider deps={dependencies} providers={providers}>
      {children}
    </ComposeProvider>
  )
}