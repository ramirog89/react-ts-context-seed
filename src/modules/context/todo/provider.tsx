import { useMemo, createContext, useReducer, useContext } from 'react';

import { GeneralModel, TodoModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';
import { useGeneral } from '../general';

interface ITodoContext {
  state: IState;
  addTodo: (todo: TodoModel.ITodo) => void;
  removeTodo: (id: string) => void;
  fetchTodoList: () => void;
}

export const TodoContext = createContext<ITodoContext | null>(null);

export const useTodo = () => {
  const context = useContext(TodoContext);
  if (!context) {
    throw new Error('TodoContext is not wrapped with TodoContext.Provider');
  }
  return context;
};

export const TodoProvider = (props: IProviderProps) => {
  const { deps } = props;
  const [state, dispatch] = useReducer(reducer, initialState);
  const { addToast } = useGeneral();

  const fetchTodoList = async () => {
    try {
      const list = await deps.apiService.getTodo();
      dispatch({ type: ActionType.ADD_TODO_LIST, payload: { list } });
    } catch (e) {
      addToast({ message: 'error', type: GeneralModel.ToastType.ERROR });
    }
  }

  const addTodo = (todo: TodoModel.ITodo) => {
    const id = new Date().toString();
    dispatch({ type: ActionType.ADD_TODO , payload: { todo: { ...todo, id } } });
    addToast({ message: 'success', type: GeneralModel.ToastType.SUCCESS });
  };

  const removeTodo = (id: string) => {
    dispatch({ type: ActionType.REMOVE_TODO , payload: { id } });
  };

  const value = useMemo(() => ({ state, addTodo, removeTodo, fetchTodoList }), [state]); // eslint-disable-line

  return <TodoContext.Provider value={value} {...props} />
};

