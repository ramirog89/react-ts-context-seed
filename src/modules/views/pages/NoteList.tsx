import { useCallback, memo, useEffect, useState } from 'react';
import { useTodo } from '../../context/todo';

const NoteList = () => {
  const { state, addTodo, removeTodo, fetchTodoList } = useTodo();
  const [isLoading, setLoading] = useState(true);

  const addTodoHandler = useCallback(() => {
    addTodo({
      title: 'some random title',
      description: `some random description ${state.todoList.length + 1}`,
      status: true
    });
  }, [state.todoList, addTodo]);

  const removeTodoHandler = useCallback((id: number) => {
    removeTodo(id);
  }, [removeTodo]);


  useEffect(() => {
    (async function() {
      await fetchTodoList();
      setLoading(false);
    })();
  }, []); // eslint-disable-line

  if (isLoading) {
    return <>Loading...</>;
  }

  return (
    <>
      <button data-testid="add-btn" onClick={addTodoHandler}>Add todo</button>
      {state.todoList.map((todo) => (
        <div data-testid="todo-item" key={todo.id}>
          {todo.description} <button data-testid="delete-btn" onClick={() => removeTodoHandler(todo.id as any)}>X</button>
        </div>
      ))}
    </>
  );
}

export default memo(NoteList);

