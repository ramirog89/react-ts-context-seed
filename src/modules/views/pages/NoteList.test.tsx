import { act, fireEvent, waitFor } from '@testing-library/react';
import { getTodo_1, getTodo_2 } from '../../../test/entities';
import { render } from '../../../test/render';

import { TodoContext } from '../../context/todo/provider';
import NoteList from './NoteList';

describe('NoteList', () => {
  let wrapper: any;

  it('should add todo', async () => {
    const mockAddTodo = jest.fn();
    const mockFetchTodoList = jest.fn().mockReturnValue([getTodo_1(), getTodo_2()]);

    await waitFor(() => {
      wrapper = render(
        <TodoContext.Provider value={{ state: { todoList: [] }, addTodo: mockAddTodo, removeTodo: jest.fn(), fetchTodoList: mockFetchTodoList }}>
          <NoteList />
        </TodoContext.Provider>
      );
    })

    act(() => {
      fireEvent.click(wrapper.getByTestId('add-btn'));
    });

    expect(mockAddTodo).toHaveBeenCalledWith({"description": "some random description 1", "status": true, "title": "some random title"});
  });


  it('should delete todo', async () => {
    const mockRemoveTodo = jest.fn();

    await waitFor(() => {
      wrapper = render(
        <TodoContext.Provider value={{ state: { todoList: [getTodo_1()] }, addTodo: jest.fn(), removeTodo: mockRemoveTodo, fetchTodoList: jest.fn() }}>
          <NoteList />
        </TodoContext.Provider>
      );
    });

    const deleteTodo = wrapper.getAllByTestId('delete-btn')[0];

    act(() => {
      fireEvent.click(deleteTodo);
    })

    expect(mockRemoveTodo).toHaveBeenCalledWith(1);
  });


  it('should render loading', () => {
    waitFor(() => {
      wrapper = render(
        <TodoContext.Provider value={{ state: { todoList: [] }, addTodo: jest.fn(), removeTodo: jest.fn(), fetchTodoList: jest.fn() }}>
          <NoteList />
        </TodoContext.Provider>
      );
    });
    expect(wrapper.container).toMatchSnapshot();
  });
});
