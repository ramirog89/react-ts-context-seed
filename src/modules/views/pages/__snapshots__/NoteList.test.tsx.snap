// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteList should render loading 1`] = `
<div>
  Loading...
</div>
`;
